deeppavlov/core/trainers/trainers.py

```diff
@@ -0,0 +1,447 @@
+        self.fit_chainer(iterator)
+        self._saved = True
+
+    def test(self, data: Iterable[Tuple[Collection[Any], Collection[Any]]],
+             metrics: Optional[Collection[Metric]] = None, *,
+             start_time: Optional[float] = None, show_examples: Optional[bool] = None) -> dict:
+
+        if start_time is None:
+            start_time = time.time()
+        if show_examples is None:
+            show_examples = self.show_examples
+        if metrics is None:
+            metrics = self.metrics
+
+        expected_outputs = list(set().union(self._chainer.out_params, *[m.inputs for m in metrics]))
+
+        outputs = {out: [] for out in expected_outputs}
+        examples = 0
+
+        data = islice(data, self.max_test_batches)
+
+        for x, y_true in data:
+            examples += len(x)
+            y_predicted = list(self._chainer.compute(list(x), list(y_true), targets=expected_outputs))
+            if len(expected_outputs) == 1:
+                y_predicted = [y_predicted]
+            for out, val in zip(outputs.values(), y_predicted):
+                out += list(val)
+
+        if examples == 0:
+            return {'eval_examples_count': 0, 'metrics': None, 'time_spent': str(datetime.timedelta(seconds=0))}
+
+        metrics_values = [(m.name, m.fn(*[outputs[i] for i in m.inputs])) for m in metrics]
+
+        report = {
+            'eval_examples_count': examples,
+            'metrics': prettify_metrics(metrics_values),
+            'time_spent': str(datetime.timedelta(seconds=round(time.time() - start_time + 0.5)))
+        }
+
+        if show_examples:
+            y_predicted = zip(*[y_predicted_group
+                                for out_name, y_predicted_group in zip(expected_outputs, y_predicted)
+                                if out_name in self._chainer.out_params])
+            if len(self._chainer.out_params) == 1:
+                y_predicted = [y_predicted_item[0] for y_predicted_item in y_predicted]
+            report['examples'] = [{
+                'x': x_item,
+                'y_predicted': y_predicted_item,
+                'y_true': y_true_item
+            } for x_item, y_predicted_item, y_true_item in zip(x, y_predicted, y_true)]
+
+        return report
+
+    def evaluate(self, iterator: DataLearningIterator, data_types: Optional[Iterable[str]] = None, *,
+                 print_reports: bool = True) -> Dict[str, dict]:
+        self._load()
+        if data_types is None:
+            data_types = self.evaluation_targets
+
+        res = {}
+
+        for data_type in data_types:
+            data_gen = iterator.gen_batches(self.batch_size, data_type=data_type, shuffle=False)
+            report = self.test(data_gen)
+            res[data_type] = report
+            if print_reports:
+                print(json.dumps({data_type: report}, ensure_ascii=False))
+
+        return res
+
+
+@register('nn_trainer')
+class NNTrainer(FitTrainer):
```

Comment:
I see an analogy between (Estimator, NNModel) and (FitTrainer, NNTrainer) abstractions. But NNTrainer inherits from FitTrainer and NNModel doesn't (from Estimator). Do you really want to use FitTrainer, not Trainer name here?